### PR TITLE
Emote leaderboard port

### PR DESCRIPTION
**Pull request by AlsoGaijutsu**
# Changes

This PR ports the emote leaderboard to live-elo. Here are the main units of change:
- New `SevenTVClient` stored in AppState. Pretty much just rips the core functionality from the old implementation, but only keeps the emote name instead of full details.
- `TwitchMessages` now store the twitch emotes from twitchirc
- `EmoteExtractingSource` decorator that parses messages sent by the inner to look for emotes. These `EmoteMessages` are stored in a queue to be sent to processors based on a tokio::select!
- `EmoteCountScoring` that gives each emote one point for each time it's seen

## Checklist

- [x] My code compiles
- [x] I have committed all the files needed to build the project (check if your file is found in `.gitignore`)
- [x] If I'm introducing a new step in the build process, I have documented / automated it
- [x] I have tested my changes (minimally with one Twitch VOD)

### CODE REVIEW

File: live-elo/src/main.rs [hunk 1]
```diff
@@ -100,12 +107,13 @@ async fn main() {
...
-                .clone()
-                .expect("need twitch channel name"),
-        ));
+        tokio_task_builder = tokio_task_builder.add_source(EmoteExtractingSource::new(
+            TwitchMessageSourceHandle::spawn(
+                GLOBAL_CONFIG
+                    .twitch_channel_name
+                    .clone()
+                    .expect("need twitch channel name"),
+        ), app_state.clone()));
```
Comment (jameshi16) [inline live-elo/src/main.rs:116]: @owobred can you take a quick look at this? I don't think this is a good approach, I feel like this should go directly into `EmoteCountSourcing` or something

Comment (AlsoGaijutsu) [reply live-elo/src/main.rs:116]: How about I can change it so `EmoteMessage` contains a count, that way the source only sends one message per unique emote in the message?

Comment (jameshi16) [reply live-elo/src/main.rs:116]: :shrug: something about wrapping the source handle to add additional features seem like it'll eventually go out of control

imagine we add more things are processed based on chat messages, do we wrap this with another extracting source?

Comment (jameshi16) [reply live-elo/src/main.rs:116]: i think more precisely, is there an environment where we'd want to process twitch messages _without_ 7tv emotes?

**Comment by jameshi16** [comment]
merged manually

**Pull request closed without merging**